rgdps/api/commands/framework.py

```diff
@@ -0,0 +1,533 @@
+from __future__ import annotations
+
+from abc import ABC
+from abc import abstractmethod
+from dataclasses import dataclass
+from enum import Enum
+from typing import Any
+from typing import Callable
+from typing import get_type_hints
+from typing import Protocol
+from typing import Type
+from typing import TYPE_CHECKING
+from typing import TypeVar
+
+from rgdps import logger
+from rgdps import repositories
+from rgdps.common.context import Context
+from rgdps.constants.users import UserPrivileges
+from rgdps.models.level import Level
+from rgdps.models.user import User
+
+if TYPE_CHECKING:
+    import httpx
+    from meilisearch_python_async import Client as MeiliClient
+    from redis.asyncio import Redis
+
+    from rgdps.models.user import User
+    from rgdps.common.cache.base import AbstractAsyncCache
+    from rgdps.services.mysql import AbstractMySQLService
+    from rgdps.services.storage import AbstractStorage
+
+
+# Private parsing functions.
+_CASTABLE = [
+    str,
+    int,
+    float,
+]
+SupportedTypes = str | int | float | bool | Level | User | Enum
+
+
+async def _level_by_ref(ctx: CommandContext, ref_value: str) -> Level:
+    if ref_value.isnumeric():
+        res = await repositories.level.from_id(
+            ctx,
+            int(ref_value),
+        )
+    else:
+        res = await repositories.level.from_name(
+            ctx,
+            ref_value,
+        )
+
+    if res is None:
+        raise ValueError("Could not find level from the given reference!")
+
+    return res
+
+
+async def _user_by_ref(ctx: CommandContext, ref_value: str) -> User:
+    if ref_value.isnumeric():
+        res = await repositories.user.from_id(
+            ctx,
+            int(ref_value),
+        )
+
+    else:
+        res = await repositories.user.from_name(
+            ctx,
+            ref_value,
+        )
+
+    if res is None:
+        raise ValueError("Could not find user from the given reference!")
+
+    return res
+
+
+T = TypeVar("T")
+
+
+async def _parse_to_type(ctx: CommandContext, value: str, cast: Type[T]) -> T:
```

Comment:
can use the builtin generic since 3.9
```suggestion
async def _parse_to_type(ctx: CommandContext, value: str, cast: type[T]) -> T:
```